homeassistant/components/derivative/sensor.py

```diff
@@ -0,0 +1,180 @@
+from homeassistant.const import (
+    CONF_NAME,
+    ATTR_UNIT_OF_MEASUREMENT,
+    STATE_UNKNOWN,
+    STATE_UNAVAILABLE,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.event import async_track_state_change
+from homeassistant.helpers.restore_state import RestoreEntity
+
+
+# mypy: allow-untyped-defs, no-check-untyped-defs
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_SOURCE_ID = "source"
+
+CONF_SOURCE_SENSOR = "source"
+CONF_ROUND_DIGITS = "round"
+CONF_UNIT_PREFIX = "unit_prefix"
+CONF_UNIT_TIME = "unit_time"
+CONF_UNIT_OF_MEASUREMENT = "unit"
+
+# SI Metric prefixes
+UNIT_PREFIXES = {None: 1, "k": 10 ** 3, "G": 10 ** 6, "T": 10 ** 9}
+
+# SI Time prefixes
+UNIT_TIME = {"s": 1, "min": 60, "h": 60 * 60, "d": 24 * 60 * 60}
+
+ICON = "mdi:chart-line"
+
+DEFAULT_ROUND = 3
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Required(CONF_SOURCE_SENSOR): cv.entity_id,
+        vol.Optional(CONF_ROUND_DIGITS, default=DEFAULT_ROUND): vol.Coerce(int),
+        vol.Optional(CONF_UNIT_PREFIX, default=None): vol.In(UNIT_PREFIXES),
+        vol.Optional(CONF_UNIT_TIME, default="h"): vol.In(UNIT_TIME),
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string,
+    }
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the derivative sensor."""
+    derivative = DerivativeSensor(
+        config[CONF_SOURCE_SENSOR],
+        config.get(CONF_NAME),
+        config[CONF_ROUND_DIGITS],
+        config[CONF_UNIT_PREFIX],
+        config[CONF_UNIT_TIME],
+        config.get(CONF_UNIT_OF_MEASUREMENT),
+    )
+
+    async_add_entities([derivative])
+
+
+class DerivativeSensor(RestoreEntity):
+    """Representation of an derivative sensor."""
+
+    def __init__(
+        self,
+        source_entity,
+        name,
+        round_digits,
+        unit_prefix,
+        unit_time,
+        unit_of_measurement,
+    ):
+        """Initialize the derivative sensor."""
+        _LOGGER.warning("unit_of_measurement: %s", unit_of_measurement)
+        self._sensor_source_id = source_entity
+        self._round_digits = round_digits
+        self._state = 0
+
+        self._name = name if name is not None else f"{source_entity} derivative"
+
+        if unit_of_measurement is None:
+            self._unit_template = (
+                f"{'' if unit_prefix is None else unit_prefix}{{}}/{unit_time}"
+            )
+            # we postpone the definition of unit_of_measurement to later
+            self._unit_of_measurement = None
+        else:
+            self._unit_of_measurement = unit_of_measurement
+
+        self._unit_prefix = UNIT_PREFIXES[unit_prefix]
+        self._unit_time = UNIT_TIME[unit_time]
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
```

Comment:
Do we need to accept zero as valid value?

Reply:
I think so, but I'm not sure. Could you elaborate why you ask it ?

Reply:
`if 0:` will evaluate as falsy. If we need to handle zero as value for state we need to make the check more specific. Eg:

```py
if state is not None:
```